v2/test/integration/mpi_job_controller_test.go

```diff
@@ -312,7 +312,7 @@ func startController(ctx context.Context, kClient kubernetes.Interface, mpiClien
 	go kubeInformerFactory.Start(ctx.Done())
 	go mpiInformerFactory.Start(ctx.Done())
 
-	go ctrl.Run(1, ctx.Done())
+	ctrl.Run(1, ctx.Done()) //nolint:errcheck
```

Comment:
Maybe, to avoid the following lint error, the only way is to disable lint.

```shell
run golangci-lint
  Running [/home/runner/golangci-lint-1.50.1-linux-amd64/golangci-lint run --out-format=github-actions --path-prefix=v2 --timeout=5m] in [/home/runner/work/mpi-operator/mpi-operator/v2] ...
  Error: Error return value of `ctrl.Run` is not checked (errcheck)
```

Reply:
this has to be in a go routine.

an alternative would be:

```golang
go func () {
  if err := ctr.Run(1, ctx.Done()); err != nil {
    t.Fatalf("Could not run controller: %s", err)
  }
}()
```

Reply:
@alculquicondor 
Previously, I tried to that solutions. Although, the controller was killed by a simple reconciling error...

```shell
I0107 00:08:57.389723   27110 mpi_job_controller.go:436] Successfully synced 'default/tensorflow-benchmarks'
--- FAIL: TestMPIJobSuccess (0.39s)
    mpi_job_controller_test.go:49: Using namespace test-w7k5z
    mpi_job_controller_test.go:114: Updating worker Pods to Running phase: Operation cannot be fulfilled on pods "job-worker-0": the object has been modified; please apply your changes to the latest version and try again
```

Reply:
The above error occurred in the case of using `go ctrl.Run(1, ctx.Done())`, too. Therefore, the other logic seems to need to be corrected. Sorry for confusing you.

Reply:
On my local, the `make test` command does not seem to work fine. Although, In CI, that command works fine. 🤔

Reply:
Done.

Reply:
what's the error you see? It was working last time I checked in my local.

Reply:
I faced the following error when I ran `make test` with `USE_EXISTING_CLUSTER=true`.

```shell
I0107 00:08:57.389723   27110 mpi_job_controller.go:436] Successfully synced 'default/tensorflow-benchmarks'
--- FAIL: TestMPIJobSuccess (0.39s)
    mpi_job_controller_test.go:49: Using namespace test-w7k5z
    mpi_job_controller_test.go:114: Updating worker Pods to Running phase: Operation cannot be fulfilled on pods "job-worker-0": the object has been modified; please apply your changes to the latest version and try again
```

https://github.com/kubeflow/mpi-operator/pull/494#discussion_r1063554169

Because of using an existing cluster, I can not use the kubebuilder envtest of v1.19 since the envtest of that version does not support my local :(
ref: https://github.com/kubeflow/katib/issues/1858

Reply:
<details>
<summary>full</summary>

```shell
$ USE_EXISTING_CLUSTER=true make test
cd v2 && go test -covermode atomic -coverprofile=profile.cov ./...
?       github.com/kubeflow/mpi-operator/v2/cmd/mpi-operator    [no test files]
?       github.com/kubeflow/mpi-operator/v2/cmd/mpi-operator/app        [no test files]
?       github.com/kubeflow/mpi-operator/v2/cmd/mpi-operator/app/options        [no test files]
ok      github.com/kubeflow/mpi-operator/v2/pkg/apis/kubeflow/v2beta1   0.270s  coverage: 50.0% of statements
ok      github.com/kubeflow/mpi-operator/v2/pkg/apis/kubeflow/validation        0.185s  coverage: 100.0% of statements
?       github.com/kubeflow/mpi-operator/v2/pkg/client/clientset/versioned      [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/clientset/versioned/fake [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/clientset/versioned/scheme       [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/clientset/versioned/typed/kubeflow/v2beta1       [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/clientset/versioned/typed/kubeflow/v2beta1/fake  [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/informers/externalversions       [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/informers/externalversions/internalinterfaces    [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/informers/externalversions/kubeflow      [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/informers/externalversions/kubeflow/v2beta1      [no test files]
?       github.com/kubeflow/mpi-operator/v2/pkg/client/listers/kubeflow/v2beta1 [no test files]
ok      github.com/kubeflow/mpi-operator/v2/pkg/controller      0.381s  coverage: 67.0% of statements
?       github.com/kubeflow/mpi-operator/v2/pkg/version [no test files]
I0107 01:41:10.210715   37869 mpi_job_controller.go:301] Setting up event handlers
I0107 01:41:10.213300   37869 mpi_job_controller.go:360] Starting MPIJob controller
I0107 01:41:10.213318   37869 mpi_job_controller.go:363] Waiting for informer caches to sync
I0107 01:41:10.314458   37869 mpi_job_controller.go:373] Starting workers
I0107 01:41:10.314486   37869 mpi_job_controller.go:379] Started workers
I0107 01:41:10.314791   37869 event.go:282] Event(v1.ObjectReference{Kind:"MPIJob", Namespace:"test-z56sp", Name:"job", UID:"cdad05c9-97ef-435b-a69f-65115448a8fa", APIVersion:"kubeflow.org/v2beta1", ResourceVersion:"854", FieldPath:""}): type: 'Normal' reason: 'MPIJobCreated' MPIJob test-z56sp/job is created.
I0107 01:41:10.435262   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (120.741792ms)
I0107 01:41:10.435281   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
I0107 01:41:10.435720   37869 event.go:282] Event(v1.ObjectReference{Kind:"MPIJob", Namespace:"test-z56sp", Name:"job", UID:"cdad05c9-97ef-435b-a69f-65115448a8fa", APIVersion:"kubeflow.org/v2beta1", ResourceVersion:"854", FieldPath:""}): type: 'Normal' reason: 'MPIJobCreated' MPIJob test-z56sp/job is created.
I0107 01:41:10.447326   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (12.035625ms)
E0107 01:41:10.447365   37869 mpi_job_controller.go:441] error syncing 'test-z56sp/job': Operation cannot be fulfilled on mpijobs.kubeflow.org "job": the object has been modified; please apply your changes to the latest version and try again
I0107 01:41:10.449020   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (1.641042ms)
I0107 01:41:10.449032   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
I0107 01:41:10.459410   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (1.500458ms)
I0107 01:41:10.459424   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
I0107 01:41:10.691257   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (14.924583ms)
I0107 01:41:10.691275   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
I0107 01:41:10.698590   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (7.307333ms)
I0107 01:41:10.698606   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
I0107 01:41:10.704669   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (2.04025ms)
I0107 01:41:10.704684   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
I0107 01:41:10.791805   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (1.530667ms)
I0107 01:41:10.791827   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
--- FAIL: TestMPIJobSuccess (0.61s)
    mpi_job_controller_test.go:49: Using namespace test-z56sp
    mpi_job_controller_test.go:133: Updating launcher Pods to Running phase: Operation cannot be fulfilled on pods "job-launcher-vq49q": the object has been modified; please apply your changes to the latest version and try again
I0107 01:41:10.799542   37869 mpi_job_controller.go:381] Shutting down workers
I0107 01:41:10.805321   37869 mpi_job_controller.go:301] Setting up event handlers
I0107 01:41:10.805449   37869 mpi_job_controller.go:360] Starting MPIJob controller
I0107 01:41:10.805455   37869 mpi_job_controller.go:363] Waiting for informer caches to sync
I0107 01:41:10.906417   37869 mpi_job_controller.go:373] Starting workers
I0107 01:41:10.906450   37869 mpi_job_controller.go:379] Started workers
I0107 01:41:10.915361   37869 mpi_job_controller.go:454] Finished syncing job "test-z56sp/job" (8.873083ms)
I0107 01:41:10.915392   37869 mpi_job_controller.go:436] Successfully synced 'test-z56sp/job'
I0107 01:41:10.915619   37869 event.go:282] Event(v1.ObjectReference{Kind:"MPIJob", Namespace:"test-4rf7s", Name:"job", UID:"ae5a2f0d-faa7-45d6-b21f-d8e8be8d8a62", APIVersion:"kubeflow.org/v2beta1", ResourceVersion:"893", FieldPath:""}): type: 'Normal' reason: 'MPIJobCreated' MPIJob test-4rf7s/job is created.
I0107 01:41:10.946721   37869 mpi_job_controller.go:454] Finished syncing job "test-4rf7s/job" (31.31275ms)
I0107 01:41:10.946752   37869 mpi_job_controller.go:436] Successfully synced 'test-4rf7s/job'
I0107 01:41:10.947016   37869 event.go:282] Event(v1.ObjectReference{Kind:"MPIJob", Namespace:"test-4rf7s", Name:"job", UID:"ae5a2f0d-faa7-45d6-b21f-d8e8be8d8a62", APIVersion:"kubeflow.org/v2beta1", ResourceVersion:"893", FieldPath:""}): type: 'Normal' reason: 'MPIJobCreated' MPIJob test-4rf7s/job is created.
I0107 01:41:10.953843   37869 mpi_job_controller.go:454] Finished syncing job "test-4rf7s/job" (7.079542ms)
E0107 01:41:10.953875   37869 mpi_job_controller.go:441] error syncing 'test-4rf7s/job': Operation cannot be fulfilled on mpijobs.kubeflow.org "job": the object has been modified; please apply your changes to the latest version and try again
I0107 01:41:10.955401   37869 mpi_job_controller.go:454] Finished syncing job "test-4rf7s/job" (1.518792ms)
I0107 01:41:10.955413   37869 mpi_job_controller.go:436] Successfully synced 'test-4rf7s/job'
I0107 01:41:10.968515   37869 mpi_job_controller.go:454] Finished syncing job "test-4rf7s/job" (3.718541ms)
I0107 01:41:10.968544   37869 mpi_job_controller.go:436] Successfully synced 'test-4rf7s/job'
I0107 01:41:11.116070   37869 mpi_job_controller.go:454] Finished syncing job "test-4rf7s/job" (17.483042ms)
I0107 01:41:11.116087   37869 mpi_job_controller.go:436] Successfully synced 'test-4rf7s/job'
--- FAIL: TestMPIJobFailure (0.33s)
    mpi_job_controller_test.go:175: Using namespace test-4rf7s
    mpi_job_controller_test.go:240: Updating launcher Pods to Running phase: Operation cannot be fulfilled on pods "job-launcher-h5q4k": StorageError: invalid object, Code: 4, Key: /registry/pods/test-4rf7s/job-launcher-h5q4k, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 113e7596-aaf0-4ffd-8d9d-177a9507a550, UID in object meta: 
FAIL
coverage: [no statements]
FAIL    github.com/kubeflow/mpi-operator/v2/test/integration    3.099s
FAIL
make: *** [test] Error 1
```

</details>

Reply:
TBH, I don't think it makes sense to run integration tests in a real cluster. The test code will be competing with kubelet to do pod updates.

Reply:
I see. So that behavior was expected. Thanks!